### PR TITLE
[libxml2] Update to 2.9.12 and use upstream CMakeLists.txt

### PR DESCRIPTION
**Pull request by Neumann-A**
- patched the library name since the configure skript will not output suffixed names and because it was easier to update it this way.

### CODE REVIEW

File: ports/libxml2/fix_cmakelist.patch [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
++++ b/libxml-2.0.pc.in
+@@ -7,7 +7,8 @@ modules=@WITH_MODULES@
+ Name: libXML
+ Version: @VERSION@
+ Description: libXML library version2.
+-Requires:
+-Libs: -L${libdir} -lxml2
+-Libs.private: @ICU_LIBS@ @THREAD_LIBS@ @Z_LIBS@ @LZMA_LIBS@ @ICONV_LIBS@ @M_LIBS@ @WIN32_EXTRA_LIBADD@ @LIBS@
++Requires: 
++Requires.private: @ICU_LIBS@ @Z_LIBS@ @LZMA_LIBS@
```
Comment (dg0yt) [inline ports/libxml2/fix_cmakelist.patch:116]: Using `Requires` instead of `Libs` :+1: 
I was already wondering if I could use this, to improve portability with less effort.

**Comment by NancyLi1013** [comment]
Hi @Neumann-A 

Thanks for updating this port.

`libxml2 `failed on x64-osx and x64-linux with the errors:
```
/Users/vagrant/Data/buildtrees/libxml2/src/b0675e461b-e0317cc2b7.clean/include/libxml/encoding.h:31:10: fatal error: 'unicode/ucnv.h' file not found
#include <unicode/ucnv.h>
```

File: ports/libxml2/fix_cmakelist.patch [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+-	DESTINATION ${CMAKE_INSTALL_LIBDIR}/cmake/libxml2-${PROJECT_VERSION}
++	DESTINATION ${CMAKE_INSTALL_LIBDIR}/cmake/libxml2
+ 	COMPONENT development
+ )
+ 
+ write_basic_package_version_file(
+ 	${CMAKE_CURRENT_BINARY_DIR}/libxml2-config-version.cmake
+ 	VERSION ${PROJECT_VERSION}
+-	COMPATIBILITY ExactVersion
++	COMPATIBILITY SameMinorVersion
```
Comment (Neumann-A) [inline ports/libxml2/fix_cmakelist.patch:87]: change to `SameMajorVersion`? The pc file (-2.0) seems to suggest that.

File: ports/libxml2/vcpkg.json [hunk 3]
```diff
@@ -1,10 +1,10 @@
 {
   "name": "libxml2",
-  "version-semver": "2.9.10",
-  "port-version": 6,
+  "version-semver": "2.9.12",
   "description": "Libxml2 is the XML C parser and toolkit developed for the Gnome project (but usable outside of the Gnome platform).",
   "homepage": "https://xmlsoft.org/",
   "dependencies": [
+    "icu",
```
Comment (Neumann-A) [inline ports/libxml2/vcpkg.json:7]:
```suggestion
```
ICU stuff should be a separate PR.

**Comment by dg0yt** [comment]
Related: #15390 @JackBoosY

**Review by NancyLi1013 — APPROVED** [review]

**Review by JackBoosY — APPROVED** [review]

**Comment by clemenswasser** [comment]
Hi @Neumann-A,
This PR only partially works for me, I get the following Error when trying to use LibXml2:
```
CMake Error at <VCPKG_ROOT>/installed/x64-windows-static/share/libxml2/vcpkg-cmake-wrapper.cmake:39 (target_link_libraries):
  Cannot specify link libraries for target "LibXml2::LibXml2" which is not
  built by this project.
Call Stack (most recent call first):
  <VCPKG_ROOT>/scripts/buildsystems/vcpkg.cmake:815 (include)
  llvm-project/clang/CMakeLists.txt:207 (find_package)
```
I am experiencing similar issues for Iconv: https://github.com/microsoft/vcpkg/pull/15390#issuecomment-826304603

Since nobody else is experiencing these issues. Am I using the wrong CMake version (3.20.2) or am I missing a required CMake configuration?

**Comment by Neumann-A** [comment]
> Since nobody else is experiencing these issues. Am I using the wrong CMake version (3.20.2) or am I missing a required CMake configuration?

check if you are really using 3.20.2. looks more like < 3.17 or whatever version target_link_libraries was allowed for imported targets.

**Comment by clemenswasser** [comment]
No I am definitely using Cmake 3.20.2
```
PS> cmake --version
cmake version 3.20.2

CMake suite maintained and supported by Kitware (kitware.com/cmake).
```

**Comment by JackBoosY** [comment]
@clemenswasser Since there is a bug of cmake, please use cmake 3.21.0 or later.

**Comment by clemenswasser** [comment]
> @clemenswasser Since there is a bug of cmake, please use cmake 3.21.0 or later.

But CMake 3.21.0 isn't even released?

**Comment by Neumann-A** [comment]
@clemenswasser Please print `${CMAKE_VERSION}` in the script you are using. If it were a bug in 3.20.2 we should also see it in CI shouldn't we?

**Comment by Neumann-A** [comment]
also make sure that your script has a call to `cmake_minimum_required` otherwise strange things could happen

**Comment by clemenswasser** [comment]
Sorry totally my fault. The script was using `cmake_minimum_required(VERSION 3.13.4)`.
But can't you set the needed policies or even `cmake_minimum_required` in your `vcpkg-cmake-wrapper.cmake`?

**Comment by clemenswasser** [comment]
Is there a way I can dump out all of the set cmake-policies and the currently set compatibility version?

**Comment by JackBoosY** [comment]
Depends on https://github.com/microsoft/vcpkg/pull/17983.

**Comment by JackBoosY** [comment]
Please update the version using `x-add-version --overwrite-version`.

**Review by JackBoosY — APPROVED** [review]

**Comment by JackBoosY** [comment]
I'm curious why you didn't encounter iconv search problem in linux, see details in https://gitlab.kitware.com/cmake/cmake/-/issues/22114.

**Comment by clemenswasser** [comment]
I found a fix for my problem. The issue seems to be that `find_package` gets called multiple times and that the `LibXml2::LibXml2` target is immutable at that point. To solve this I just putting `if(NOT DEFINED LibXml2_FOUND)` around the whole code in `vcpkg-cmake-wrapper.cmake`.

**Comment by JackBoosY** [comment]
@Neumann-A Can you consider accepting @clemenswasser 's suggestion?

**Comment by clemenswasser** [comment]
LibXml2 should also link with `Ws2_32.lib` on Windows.
This is currently missing and you get linker errors.
It is required in `nanohttp.c`

**Comment by Neumann-A** [comment]
@JackBoosY: I cannot prohibit a `find_package` call if `LibXml2_FOUND` is already found due to the possibility that it is called with different `COMPONENTS` (although that does not make sense here but it is a CMake rule). Also without a `--trace/--trace-expand` log with the problem I am not going to change that (Tipp: Locally merge #18032 to get a readable log). I suspect a target promotion to `GLOBAL` to be the problem here. 

> LibXml2 should also link with Ws2_32.lib on Windows.

It already does in all three modes (CONFIG, MODULE & PC)

**Comment by clemenswasser** [comment]
lzma is also missing on the target

File: ports/libxml2/vcpkg-cmake-wrapper.cmake [hunk 4]
```diff
@@ -1,24 +1,53 @@
 _find_package(${ARGS})
 if(LibXml2_FOUND)
+    list(APPEND LIBXML2_INCLUDE_DIRS "${_VCPKG_INSTALLED_DIR}/${VCPKG_TARGET_TRIPLET}/include")
+    list(APPEND LIBXML2_INCLUDE_DIR "${_VCPKG_INSTALLED_DIR}/${VCPKG_TARGET_TRIPLET}/include") # This is wrong but downstream doesn't correctly use _DIR vs _DIRS variables
+    if(TARGET LibXml2::LibXml2)
+       target_include_directories(LibXml2::LibXml2 INTERFACE "${_VCPKG_INSTALLED_DIR}/${VCPKG_TARGET_TRIPLET}/include")
+    endif()
+endif()
+if(LibXml2_FOUND AND @VCPKG_LIBRARY_LINKAGE@ STREQUAL "static" AND NOT ${ARGV0}_CONFIG)
```
Comment (ras0219-msft) [inline ports/libxml2/vcpkg-cmake-wrapper.cmake:9]:
```suggestion
if(LibXml2_FOUND AND "@VCPKG_LIBRARY_LINKAGE@" STREQUAL "static" AND NOT ${ARGV0}_CONFIG)
```

File: ports/libxml2/vcpkg-cmake-wrapper.cmake [hunk 5]
```diff
@@ -1,24 +1,53 @@
...
+        if(TARGET LibXml2::LibXml2)
+            target_link_libraries(LibXml2::LibXml2 INTERFACE "ws2_32")
+        endif()
     endif()
     if(ICONV_LIBRARIES)
         list(APPEND LIBXML2_LIBRARIES ${ICONV_LIBRARIES})
         if(TARGET LibXml2::LibXml2)
-           target_link_libraries(LibXml2::LibXml2 INTERFACE ${ICONV_LIBRARIES} )
+           target_link_libraries(LibXml2::LibXml2 INTERFACE ${ICONV_LIBRARIES})
+           target_include_directories(LibXml2::LibXml2 INTERFACE "${_VCPKG_INSTALLED_DIR}/${VCPKG_TARGET_TRIPLET}/include" )
```
Comment (ras0219-msft) [inline ports/libxml2/vcpkg-cmake-wrapper.cmake:50]:
```suggestion
```
This is handled above

File: ports/libxml2/portfile.cmake [hunk 6]
```diff
@@ -1,50 +1,95 @@
...
     PREFER_NINJA
     OPTIONS
-        -DPORT_DIR=${CMAKE_CURRENT_LIST_DIR}
-        -DWITH_HTTP=${ENABLE_NETWORK}
-        -DWITH_FTP=${ENABLE_NETWORK}
-    OPTIONS_DEBUG
-        -DINSTALL_HEADERS=OFF
+        ${FEATURE_OPTIONS}
+        -DLIBXML2_WITH_TESTS=OFF
+        #-DPORT_DIR=${CMAKE_CURRENT_LIST_DIR}
```
Comment (ras0219-msft) [inline ports/libxml2/portfile.cmake:29]:
```suggestion
```

File: ports/libxml2/vcpkg-cmake-wrapper.cmake [hunk 7]
```diff
@@ -1,24 +1,53 @@
 _find_package(${ARGS})
```
Comment (ras0219-msft) [inline ports/libxml2/vcpkg-cmake-wrapper.cmake:1]: Now that we're using upstream's CMake which has a `-config`, it would be better to reorient this file to always pull in the config mode file but then to add compatibility variables for the built-in module (`LIBXML2_INCLUDE_DIRS`, etc).

I don't think we have to do this now, but that's the direction we should move.

File: ports/libxml2/fix_cmakelist.patch [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+ if(LIBXML2_WITH_THREADS)
+@@ -483,7 +483,7 @@ endif()
+ 
+ if(LIBXML2_WITH_ZLIB)
+ 	target_link_libraries(LibXml2 PRIVATE ZLIB::ZLIB)
+-	set(Z_LIBS "-lz")
++	set(Z_LIBS "zlib")
+ endif()
+ 
+ set_target_properties(
```
Comment (ras0219-msft) [inline ports/libxml2/fix_cmakelist.patch:40]:
```suggestion
```
Since the pkgconfig file doesn't need to embed the libraries and we aren't using ICU yet, this section of the patch is not needed.

File: ports/libxml2/fix_cmakelist.patch [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
++++ b/libxml-2.0.pc.in
+@@ -7,7 +7,8 @@ modules=@WITH_MODULES@
+ Name: libXML
+ Version: @VERSION@
+ Description: libXML library version2.
+-Requires:
+-Libs: -L${libdir} -lxml2
+-Libs.private: @ICU_LIBS@ @THREAD_LIBS@ @Z_LIBS@ @LZMA_LIBS@ @ICONV_LIBS@ @M_LIBS@ @WIN32_EXTRA_LIBADD@ @LIBS@
++Requires: 
++Requires.private: @ICU_LIBS@ @Z_LIBS@ @LZMA_LIBS@
```
[the remark this answers is not in the document]
Comment (Neumann-A) [reply ports/libxml2/fix_cmakelist.patch:116]: It does. I was just to lazy to rename the variables. So I use Z_LIBS to pass zlib and LZMA_LIBS to pass liblzma (and ICU_LIBS to pass the icu pc files if somebody someday adds the feature.).

File: ports/libxml2/vcpkg-cmake-wrapper.cmake (hunk 7 above)
Comment (Neumann-A) [reply ports/libxml2/vcpkg-cmake-wrapper.cmake:1]: I vote against this. You never know if a project has a custom `Find<Something>.cmake`. Always redirecting to config breaks those custom modules.

File: ports/libxml2/fix_cmakelist.patch (hunk 8 above)
Comment (Neumann-A) [reply ports/libxml2/fix_cmakelist.patch:40]: as said above i was just to lazy to rename the variables to `something_PC_NAME` instead. Furthermore I am against hardcoding the PC names since somebody might add features to libxml2 to enable/disable zlib/lzma/icu. With the current patch no further modification will be required and only the portfile needs to be touched.

File: ports/libxml2/vcpkg-cmake-wrapper.cmake (hunk 7 above)
Comment (dg0yt) [reply ports/libxml2/vcpkg-cmake-wrapper.cmake:1]: > You never know if a project has a custom Find<Something>.cmake. Always redirecting to config breaks those custom modules.

I don't want to vote but I think I can give a recent example:
- zstd now provides `zstd::libzstd_static` and similar. [!]
- There are find modules setting variable like in folly: https://github.com/facebook/folly/blob/master/CMake/FindZstd.cmake
- There are find modules creating targets, e.g. `ZSTD::ZSTD` in tiff: https://gitlab.com/libtiff/libtiff/-/blob/master/cmake/FindZSTD.cmake

Comment (JackBoosY) [reply ports/libxml2/vcpkg-cmake-wrapper.cmake:1]: Ping @ras0219 @ras0219-msft for response.

File: ports/libxml2/vcpkg.json (hunk 3 above)
Comment (ras0219-msft) [reply ports/libxml2/vcpkg.json:7]: Should this be applied since ICU is disabled during the build?

**Review by ras0219-msft — APPROVED** [review]
LGTM once we figure out whether the `"icu"` dependency should go -- it's a heavy dependency that I'd like to avoid if possible.

**Comment by JackBoosY** [comment]
LGTM.

File: scripts/ci.baseline.txt [hunk 10]
```diff
@@ -820,6 +820,8 @@ libwandio:x64-windows-static-md=fail
 libwandio:x64-uwp=fail
 libwandio:arm64-windows=fail
 libwandio:arm-uwp=fail
+libxml2:arm-uwp=fail
+libxml2:x64-uwp=fail
```
Comment (NancyLi1013) [inline scripts/ci.baseline.txt:824]: Why set these triplets as fail here? Does `libxml2 `not support uwp? Or something else? Please help make a clarification.

Comment (Neumann-A) [reply scripts/ci.baseline.txt:824]: The error is:
`E:\vcpkg_folders\libxml2\buildtrees\libxml2\src\b0675e461b-7588e7e3fb.clean\valid.c(5243,14): error C4996: 'strcat': This function or variable may be unsafe. Consider using strcat_s instead. To disable deprecation, use _CRT_SECURE_NO_WARNINGS.`
which seems to be a fatal error on UWP. On normal windows this is always just a warning. 

According to:
https://docs.microsoft.com/en-us/cpp/cppcx/crt-functions-not-supported-in-universal-windows-platform-apps?view=msvc-160
these functions are available, so I assume setting _CRT_SECURE_NO_WARNINGS would get rid of the error? 
(But that is not something I want to fix/investigate in this PR)

Comment (Neumann-A) [reply scripts/ci.baseline.txt]: after looking at all the errors and the UWP available functions I came to the conclusion that it should not support UWP

File: ports/libxml2/vcpkg-cmake-wrapper.cmake [hunk 11]
```diff
@@ -1,24 +1,52 @@
...
```
Comment (JackBoosY) [inline ports/libxml2/vcpkg-cmake-wrapper.cmake:1]: Since it may call `FindLibxml2.cmake`, should we force change this to `_find_package(libxml2 CONFIG REQUIRED)`?

Comment (Neumann-A) [reply ports/libxml2/vcpkg-cmake-wrapper.cmake:1]: same question as https://github.com/microsoft/vcpkg/pull/17945#discussion_r638346809 and the answer to that is still no.

Comment (JackBoosY) [reply ports/libxml2/vcpkg-cmake-wrapper.cmake:1]: Since our wrapper provides all the macro that FindLibxml2.cmake provided, I think we can smoothly switch from FindLibxml2.cmake to libxml2-config.cmake.

Comment (Neumann-A) [reply ports/libxml2/vcpkg-cmake-wrapper.cmake:1]: Still no..... 
see 
https://github.com/microsoft/vcpkg/pull/17945#discussion_r642674529
https://github.com/microsoft/vcpkg/pull/17945#discussion_r642683804

Comment (JackBoosY) [reply ports/libxml2/vcpkg-cmake-wrapper.cmake:1]: Okay, fine.

**Comment by JackBoosY** [comment]
Ping @ras0219 @ras0219-msft for merge this PR.

**Comment by ras0219-msft** [comment]
I'll handle the merge conflicts

**Review by ras0219-msft — APPROVED** [review]

**Comment by BillyONeal** [comment]
Thanks for your contribution!

File: ports/libxml2/vcpkg.json [hunk 12]
```diff
@@ -1,12 +1,17 @@
 {
   "name": "libxml2",
-  "version-semver": "2.9.10",
-  "port-version": 8,
+  "version-semver": "2.9.12",
   "description": "Libxml2 is the XML C parser and toolkit developed for the Gnome project (but usable outside of the Gnome platform).",
   "homepage": "https://xmlsoft.org/",
+  "supports": "!uwp",
```
Comment (cenit) [inline ports/libxml2/vcpkg.json:6]: @NancyLi1013 here it is why opencv is now a cascade failure on uwp

may i ask why such a modification was merged without any discussion? many ports depend on libxml2 and this made all of them silently untested……
@ras0219-msft @BillyONeal

Comment (NancyLi1013) [reply ports/libxml2/vcpkg.json:6]: Thanks for pointing this out. @cenit 

Here is the discussion about the failure on uwp.
https://github.com/microsoft/vcpkg/pull/17945#discussion_r652548717

Comment (NancyLi1013) [reply ports/libxml2/vcpkg.json:6]: Thanks for pointing this out. @cenit 

Here is the discussion about the failure on uwp.
https://github.com/microsoft/vcpkg/pull/17945#discussion_r652548717

File: scripts/ci.baseline.txt (hunk 10 above)
Comment (cenit) [reply scripts/ci.baseline.txt]: it was supporting UWP before without any problem

Comment (cenit) [reply scripts/ci.baseline.txt]: and is now silently not testing many ports in uwp due to this

File: ports/libxml2/vcpkg.json (hunk 12 above)
Comment (NancyLi1013) [reply ports/libxml2/vcpkg.json:6]: I have added a patch to fix uwp #19221.
